aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -8,6 +8,34 @@
 
 namespace at { namespace native {
 
+
+typedef struct {
+  int err;
+  int64_t index;
+  int64_t axis;
+  int64_t size;
+} status_t;
+
+static status_t* make_device_status() {
+  cudaError_t err;
+  status_t* status;
+  AT_CUDA_CHECK(cudaMalloc(&status, sizeof *status));
+  err = cudaMemset(status, 0, sizeof *status);
+  if (err != cudaSuccess) {
+    AT_CUDA_CHECK(cudaFree(status));
+    AT_CUDA_CHECK(err);
+    return NULL; // Not reached.
+  }
+  return status;
+}
+
+static void move_status_to_host(status_t *host_status, status_t *device_status) {
+  cudaError_t err;
+  err = cudaMemcpy(host_status, device_status, sizeof *host_status, cudaMemcpyDeviceToHost);
+  AT_CUDA_CHECK(cudaFree(device_status));
```

Comment:
This will imply a sync point in every call to `gpu_index_kernel`, which will not be great for performance reasons.
We used to have `THCudaMemCpy` and `THCudaFree`, but things have changed with the caching allocator and I'm not sure what's the recommended way anymore.